File: src/eYcel/decrypt.py
"""
Excel decryption pipeline.

Workflow
--------
1. Load and validate rules.yaml.
2. Open the encrypted workbook (data_only=False).
3. Extract the global inverse text map (fake→original).
4. For each sheet, extract formulas, reverse-transform ALL cells,
   reinject formulas with text literals reversed.
5. Save the restored workbook.
"""
from pathlib import Path
from typing import Any, Dict, List

import openpyxl
from openpyxl.utils import get_column_letter


from .formula_handler import extract_formulas, clear_formula_cells
from .transformations import (
    reverse_offset_date,
    reverse_offset_number,
    reverse_scale,
    reverse_shuffle,
    transform_keep,
    reverse_text_in_formula,
)
from .yaml_handler import load_rules


# ---------------------------------------------------------------------------
# Reverse transform dispatcher
# ---------------------------------------------------------------------------

def _reverse_cell(value: Any, config: Dict[str, Any], inverse_text_map: Dict[str, str]) -> Any:
    """Apply the correct reverse transform to a single cell value."""
    # Text cells that exist in the global inverse map are always reversed,
    # regardless of column transform. This restores headers, titles, etc.
    if isinstance(value, str) and not value.startswith("="):
        if value in inverse_text_map:
            return inverse_text_map[value]

    t = config.get("transform", "keep")

    if t == "hash":
        return value  # one-way

    if t == "offset":
        if hasattr(value, "year"):
            return reverse_offset_date(value, config.get("offset_days", 0))
        return reverse_offset_number(float(value), config.get("offset", 0.0))

    if t == "scale":
        return reverse_scale(float(value), config["factor"])

    if t == "shuffle":
        return reverse_shuffle(str(value), config.get("mapping", {}))

    if t in ("anonymize", "keep"):
        return transform_keep(value)

    return value


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def load_and_validate_rules(rules_path: str) -> Dict[str, Any]:
    """Load a rules YAML file and raise on invalid content."""
    return load_rules(rules_path)


def apply_reverse_transform(
    column_data: List[Any],
    transform_config: Dict[str, Any],
) -> List[Any]:
    """Apply a reverse transform to an entire column's data list."""
    inverse = {}  # no global map needed for this legacy API
    return [
        _reverse_cell(v, transform_config, inverse) if v is not None else v
        for v in column_data
    ]


def decrypt_excel(
    encrypted_path: str,
    rules_path: str,
    output_path: str,
) -> None:
    """
    Restore an encrypted Excel file to its original values.

    Handles the global text substitution map for consistent reversal.
    """
    # ── 1. Load & validate rules ─────────────────────────────────────────────
    rules = load_and_validate_rules(rules_path)
    column_configs: Dict[str, Dict[str, Any]] = rules.get("columns", {})

    # ── 2. Extract global inverse text map ───────────────────────────────────
    global_entry = column_configs.get("__global_text_map", {})
    inverse_text_map: Dict[str, str] = global_entry.get("mapping", {})

    # ── 3. Open encrypted workbook ───────────────────────────────────────────
    src = Path(encrypted_path)
    if not src.exists():
        raise FileNotFoundError(f"Encrypted file not found: {encrypted_path}")

    wb = openpyxl.load_workbook(src, data_only=False)

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        max_row = ws.max_row or 1
        max_col = ws.max_column or 1

        if max_row < 1:
            continue

        # ── 4. Extract formulas ──────────────────────────────────────────────
        formula_map = extract_formulas(ws)
        clear_formula_cells(ws, formula_map)

        # ── 5. Reverse-transform ALL cells ───────────────────────────────────
        for col_idx in range(1, max_col + 1):
            col_letter = get_column_letter(col_idx)
            col_key = f"{sheet_name}!{col_letter}"

            cfg = column_configs.get(col_key)
            if cfg is None:
                continue
            if cfg.get("transform") == "substitute" and col_key == "__global_text_map":
                continue

            for row in range(1, max_row + 1):
                cell = ws.cell(row=row, column=col_idx)
                if cell.value is None:
                    continue
                try:
                    cell.value = _reverse_cell(cell.value, cfg, inverse_text_map)
                except Exception:
                    pass

        # ── 6. Reinject formulas with text reversed ──────────────────────────
        for (row, col), formula in formula_map.items():
            restored = reverse_text_in_formula(formula, inverse_text_map)
            ws.cell(row=row, column=col).value = restored

    # ── 7. Save restored workbook ────────────────────────────────────────────
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    wb.save(output_path)

File: eYcel-wasm/src/decrypt.rs
use crate::cell_value::CellValue;
use crate::errors::EYcelError;
use crate::formula_handler::{clear_formula_cells, extract_formulas};
use crate::transformations::*;
use crate::xlsx_io::{col_index_to_letter, read_file, write_csv, write_xlsx};
use crate::yaml_handler::{rules_from_yaml, ColumnConfig, Rules};

/// Apply the correct reverse transform to a single cell value.
/// `inverse_text_map` is the global fake→original mapping for "substitute" transforms.
fn reverse_cell(
    value: CellValue,
    config: &ColumnConfig,
    inverse_text_map: &std::collections::HashMap<String, String>,
) -> CellValue {
    // Text cells that exist in the global inverse map are always reversed.
    if let CellValue::String(ref s) = value {
        if let Some(original) = inverse_text_map.get(s) {
            return CellValue::String(original.clone());
        }
        // Fall through to column-specific transform (e.g. shuffle)
    }

    match config.transform.as_str() {
        "hash" | "anonymize" => {
            // One-way transforms — cannot reverse, return as-is
            value
        }
        "offset" => {
            if value.has_year() {
                match value {
                    CellValue::Date(d) => {
                        CellValue::Date(reverse_offset_date(d, config.offset_days.unwrap_or(0)))
                    }
                    CellValue::DateTime(dt) => CellValue::DateTime(reverse_offset_datetime(
                        dt,
                        config.offset_days.unwrap_or(0),
                    )),
                    _ => value,
                }
            } else if let Some(f) = value.as_f64() {
                CellValue::Float(reverse_offset_number(f, config.offset.unwrap_or(0.0)))
            } else {
                value
            }
        }
        "scale" => {
            if let Some(f) = value.as_f64() {
                match reverse_scale(f, config.factor.unwrap_or(1.0)) {
                    Ok(v) => CellValue::Float(v),
                    Err(_) => value,
                }
            } else {
                value
            }
        }
        "shuffle" => {
            if let Some(mapping) = &config.mapping {
                CellValue::String(reverse_shuffle(&value.to_string_repr(), mapping))
            } else {
                value
            }
        }
        _ => value, // "keep" or unknown
    }
}

/// Build a column-letter → column index map for a given sheet.
/// Rules use "SheetName!ColLetter" as keys, so we build both forms.
fn build_col_key_map(sheet_name: &str, num_cols: usize) -> std::collections::HashMap<String, usize> {
    let mut map = std::collections::HashMap::new();
    for idx in 0..num_cols {
        let letter = col_index_to_letter(idx);
        // Match "Sheet1!A" format (primary)
        map.insert(format!("{}!{}", sheet_name, letter), idx);
        // Also match bare "A" format (fallback for single-sheet files)
        map.insert(letter, idx);
    }
    map
}

/// Decrypt an encrypted file using rules.
///
/// Accepts xlsx, xls, or csv input (detected from filename).
/// Returns decrypted xlsx bytes.
///
/// **Formula guarantee:** all formulas are extracted before reverse-transformation
/// and reinjected after, so they are preserved 100%.
pub fn decrypt_excel(encrypted_bytes: &[u8], rules_yaml: &str, filename: &str) -> Result<Vec<u8>, EYcelError> {
    let sheets = decrypt_to_sheets(encrypted_bytes, rules_yaml, filename)?;
    write_xlsx(&sheets)
}

/// Decrypt and return as CSV bytes (first sheet only).
/// Accepts xlsx, xls, or csv input (detected from filename).
pub fn decrypt_excel_as_csv(encrypted_bytes: &[u8], rules_yaml: &str, filename: &str) -> Result<Vec<u8>, EYcelError> {
    let sheets = decrypt_to_sheets(encrypted_bytes, rules_yaml, filename)?;
    write_csv(&sheets)
}

/// Core decryption logic — returns decrypted SheetData.
fn decrypt_to_sheets(
    encrypted_bytes: &[u8],
    rules_yaml: &str,
    filename: &str,
) -> Result<Vec<crate::xlsx_io::SheetData>, EYcelError> {
    // 1. Parse and validate rules
    let rules: Rules =
        rules_from_yaml(rules_yaml).map_err(|e| EYcelError::Decryption(e))?;

    // 2. Extract the global inverse text map (fake→original) if present
    let inverse_text_map = rules
        .columns
        .get("__global_text_map")
        .and_then(|cfg| cfg.mapping.clone())
        .unwrap_or_default();

    // 3. Read encrypted file — auto-detect format from filename
    let mut sheets = read_file(encrypted_bytes, filename)?;

    for sheet in &mut sheets {
        if sheet.grid.is_empty() {
            continue;
        }

        let col_map = build_col_key_map(&sheet.name, sheet.num_cols());

        // 4. Extract formulas BEFORE any reverse transformation
        let formula_map = extract_formulas(&sheet.grid);

        // 5. Clear formula cells
        clear_formula_cells(&mut sheet.grid, &formula_map);

        // 6. Reverse-transform ALL cells in ALL rows (no row is special)
        for (col_key, cfg) in &rules.columns {
            if col_key == "__global_text_map" {
                continue;
            }
            if let Some(&col_idx) = col_map.get(col_key) {
                for row_idx in 0..sheet.grid.len() {
                    if col_idx < sheet.grid[row_idx].len() {
                        let cell = &sheet.grid[row_idx][col_idx];
                        if cell.is_empty() {
                            continue;
                        }
                        let restored = reverse_cell(cell.clone(), cfg, &inverse_text_map);
                        sheet.grid[row_idx][col_idx] = restored;
                    }
                }
            }
        }

        // 7. Reinject formulas with text substitutions reversed
        for (&(row, col), formula) in &formula_map {
            if row < sheet.grid.len() && col < sheet.grid[row].len() {
                let restored_formula = reverse_text_in_formula(formula, &inverse_text_map);
                sheet.grid[row][col] = CellValue::Formula(restored_formula);
            }
        }
    }

    Ok(sheets)
}

/// Replace fake text back to original inside formula strings.
fn reverse_text_in_formula(formula: &str, inverse_map: &std::collections::HashMap<String, String>) -> String {
    let mut result = String::with_capacity(formula.len());
    let chars: Vec<char> = formula.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        if chars[i] == '"' {
            let start = i + 1;
            i += 1;
            while i < chars.len() && chars[i] != '"' {
                i += 1;
            }
            let literal: String = chars[start..i].iter().collect();

            result.push('"');
            if let Some(original) = inverse_map.get(&literal) {
                result.push_str(original);
            } else {
                result.push_str(&literal);
            }
            result.push('"');

            if i < chars.len() {
                i += 1;
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }

    result
}

/// Validate rules YAML and return structured info.
/// Used by the web UI's "Validate" tab.
pub fn validate_rules_yaml(rules_yaml: &str) -> Result<Rules, String> {
    rules_from_yaml(rules_yaml)
}
